File: src/transform/flip.rs
use super::super::{
    utilities::prelude::*,
    ray::Ray,
    aabb::AABB,
    transform::prelude::*,
};

#[derive(Clone)]
pub struct Flip {

}

impl Flip {
    pub fn new() -> Self {
        Self { }
    }
}

impl Transform for Flip {
    fn transform_vector(&self, v: Vec3) -> Vec3 {
        -v
    }

    fn inverse_transform_vector(&self, v: Vec3) -> Vec3 {
        -v
    }

    fn transform_ray(&self, ray: &Ray) -> Ray {
        ray.clone()
    }

    fn inverse_transform_ray(&self, ray: &Ray) -> Ray {
        ray.clone()
    }

    fn transform_box(&self, bounding_box: &AABB) -> AABB {
        bounding_box.clone()
    }

    fn inverse_transform_box(&self, bounding_box: &AABB) -> AABB {
        bounding_box.clone()
    }
}

File: lib/Transform/Flip.cpp
#include "Transform/Flip.hpp"

auto Flip::transform_vector(const Vec3 &p) const -> Vec3 { return -p; }
auto Flip::inverse_transform_vector(const Vec3 &p) const -> Vec3 { return -p; }

auto Flip::transform_ray(const Ray &ray) const -> Ray { return ray; }
auto Flip::inverse_transform_ray(const Ray &ray) const -> Ray { return ray; }

auto Flip::transform_box(const AABB &box) const -> AABB { return box; }
auto Flip::inverse_transform_box(const AABB &box) const -> AABB { return box; }
